src/widget/widget.js

```diff
@@ -64,12 +69,71 @@ export default class Widget extends Plugin {
 		}
 
 		// Create model selection over widget.
-		const modelDocument = editor.document;
 		const modelElement = editor.editing.mapper.toModelElement( widgetElement );
-		const modelRange = ModelRange.createOn( modelElement );
 
-		modelDocument.enqueueChanges( ( ) => {
-			modelDocument.selection.setRanges( [ modelRange ] );
+		editor.document.enqueueChanges( ( ) => {
+			this._setSelectionOverElement( modelElement );
 		} );
 	}
+
+	/**
+	 * Handles {@link engine.view.Document#keydown keydown} events.
+	 *
+	 * @private
+	 * @param {utils.EventInfo} eventInfo
+	 * @param {envine.view.observer.DomEventData} domEventData
+	 */
+	_onKeydown( eventInfo, domEventData  ) {
+		const keyCode = domEventData.keyCode;
+
+		// Handle only delete and backspace.
+		if ( keyCode !== keyCodes.delete && keyCode !== keyCodes.backspace ) {
+			return;
+		}
+
+		const dataController = this.editor.data;
+		const modelDocument = this.editor.document;
+		const modelSelection = modelDocument.selection;
+
+		// Do nothing on non-collapsed selection.
+		if ( !modelSelection.isCollapsed ) {
+			return;
+		}
+
+		// Clone current selection to use it as a probe. We must leave default selection as it is so it can return
+		// to its current state after undo.
+		const probe = ModelSelection.createFromSelection( modelSelection );
+		const isForward = ( keyCode == keyCodes.delete );
+
+		dataController.modifySelection( probe, { direction: isForward ? 'forward' : 'backward' } );
+
+		const objectElement = isForward ? probe.focus.nodeBefore : probe.focus.nodeAfter;
+
+		if ( objectElement && modelDocument.schema.objects.has( objectElement.name ) ) {
```

Comment:
And if we're at the beginning of a text node? Let's say – there was just one letter before our initial selection? Then, will we get that text node?